x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/kerberos/KerberosRealm.java

```diff
@@ -0,0 +1,190 @@
+    private final KerberosTicketValidator kerberosTicketValidator;
+    private final ThreadPool threadPool;
+
+    public KerberosRealm(final RealmConfig config, final NativeRoleMappingStore nativeRoleMappingStore, final ThreadPool threadPool) {
+        this(config, nativeRoleMappingStore, new KerberosTicketValidator(), threadPool, null);
+    }
+
+    // pkg scoped for testing
+    KerberosRealm(final RealmConfig config, final NativeRoleMappingStore nativeRoleMappingStore,
+            final KerberosTicketValidator kerberosTicketValidator, final ThreadPool threadPool,
+            final Cache<String, User> userPrincipalNameToUserCache) {
+        super(KerberosRealmSettings.TYPE, config);
+        this.userRoleMapper = nativeRoleMappingStore;
+        this.userRoleMapper.refreshRealmOnChange(this);
+        final TimeValue ttl = KerberosRealmSettings.CACHE_TTL_SETTING.get(config.settings());
+        if (ttl.getNanos() > 0) {
+            this.userPrincipalNameToUserCache = (userPrincipalNameToUserCache == null)
+                    ? CacheBuilder.<String, User>builder()
+                            .setExpireAfterWrite(KerberosRealmSettings.CACHE_TTL_SETTING.get(config.settings()))
+                            .setMaximumWeight(KerberosRealmSettings.CACHE_MAX_USERS_SETTING.get(config.settings())).build()
+                    : userPrincipalNameToUserCache;
+        } else {
+            this.userPrincipalNameToUserCache = null;
+        }
+        this.kerberosTicketValidator = kerberosTicketValidator;
+        this.threadPool = threadPool;
+    }
+
+    @Override
+    public Map<String, List<String>> getAuthenticationFailureHeaders() {
+        return Collections.singletonMap(WWW_AUTHENTICATE, Collections.singletonList(NEGOTIATE_SCHEME_NAME));
+    }
+
+    @Override
+    public void expire(final String username) {
+        if (userPrincipalNameToUserCache != null) {
+            userPrincipalNameToUserCache.invalidate(username);
+        }
+    }
+
+    @Override
+    public void expireAll() {
+        if (userPrincipalNameToUserCache != null) {
+            userPrincipalNameToUserCache.invalidateAll();
+        }
+    }
+
+    @Override
+    public boolean supports(final AuthenticationToken token) {
+        return token instanceof KerberosAuthenticationToken;
+    }
+
+    @Override
+    public AuthenticationToken token(final ThreadContext context) {
+        return KerberosAuthenticationToken.extractToken(context.getHeader(AUTH_HEADER));
+    }
+
+    @Override
+    public void authenticate(final AuthenticationToken token, final ActionListener<AuthenticationResult> listener) {
+        if (token instanceof KerberosAuthenticationToken) {
+            final KerberosAuthenticationToken kerbAuthnToken = (KerberosAuthenticationToken) token;
+            final Path keytabPath =
+                    config.env().configFile().resolve(KerberosRealmSettings.HTTP_SERVICE_KEYTAB_PATH.get(config.settings()));
```

Comment:
Is there a reason to resolve this on every call to authenticate? The setting will never change, so we can resolve & store the `Path` in the constructor.